src/Middleware/ParametersNameDecodingHandler.cs

```diff
@@ -0,0 +1,53 @@
+// ------------------------------------------------------------------------------
+//  Copyright (c) Microsoft Corporation.  All Rights Reserved.  Licensed under the MIT License.  See License in the project root for license information.
+// ------------------------------------------------------------------------------
+
+using System;
+using System.Linq;
+using System.Net.Http;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Kiota.Http.HttpClientLibrary.Middleware.Options;
+
+namespace Microsoft.Kiota.Http.HttpClientLibrary.Middleware;
+
+/// <summary>
+/// This handlers decodes special characters in the request query parameters that had to be encoded due to RFC 6570 restrictions names before executing the request.
+/// </summary>
+public class ParametersNameDecodingHandler: DelegatingHandler
+{
+    /// <summary>
+    /// The options to use when decoding parameters names in URLs
+    /// </summary>
+    public ParametersNameDecodingOption EncodingOptions { get; set; }
+    /// <summary>
+    /// Constructs a new <see cref="ParametersNameDecodingHandler"/>
+    /// </summary>
+    /// <param name="options">An OPTIONAL <see cref="ParametersNameDecodingOption"/> to configure <see cref="ParametersNameDecodingHandler"/></param>
+    public ParametersNameDecodingHandler(ParametersNameDecodingOption options = default)
+    {
+        EncodingOptions = options ?? new();
+    }
+    ///<inheritdoc/>
+    protected override Task<HttpResponseMessage> SendAsync(HttpRequestMessage httpRequestMessage, CancellationToken cancellationToken)
+    {
+        if(!httpRequestMessage.RequestUri.Query.Contains('%') ||
```

Comment:
We should probably check if there is a `ParametersNameDecodingOption` set in the request message itself first to use if it exists to allow custom options on a "per request basis" like the other middleware. 

```cs
EncodingOptions = httpRequestMessage.GetRequestOption<ParametersNameDecodingOption>() ?? EncodingOptions;
```

Reply:
fixed.